csi/node_server.go

```diff
@@ -615,25 +628,33 @@ func (ns *NodeServer) getMounter(volume *longhornclient.Volume, volumeCapability
 
 	// mounter that can format and use hard coded filesystem params
 	if volumeCapability.GetMount() != nil {
-		userExt4Params, _ := ns.apiClient.Setting.ById(string(types.SettingNameMkfsExt4Parameters))
-
-		// mounter assumes ext4 by default
 		fsType := volumeCapability.GetMount().GetFsType()
 		if fsType == "" {
-			fsType = "ext4"
+			fsType = DefaultFsType
+		}
+
+		// To allow users to override the default block size,
+		// put the default block size in front of other user-defined parameters.
+		params := getBlockSizeParams(fsType)
+
+		// We allow the user to provide additional params for ext4 filesystem creation.
+		// This allows an ext4 fs to be mounted on older kernels, see https://github.com/longhorn/longhorn/issues/1208
+		if fsType == "ext4" {
+			if userParams, _ := ns.apiClient.Setting.ById(string(types.SettingNameMkfsExt4Parameters)); userParams != nil && userParams.Value != "" {
+				params += " " + userParams.Value
+			}
 		}
 
 		mounter := &mount.SafeFormatAndMount{Interface: mount.New(""), Exec: utilexec.New()}
-		// we allow the user to provide additional params for ext4 filesystem creation.
-		// this allows an ext4 fs to be mounted on older kernels, see https://github.com/longhorn/longhorn/issues/1208
-		if fsType == "ext4" && userExt4Params != nil && userExt4Params.Value != "" {
-			ext4Params := userExt4Params.Value
-			logrus.Infof("volume %v using user provided ext4 fs creation params: %s", volume.Name, ext4Params)
-			cmdParamMapping := map[string]string{"mkfs." + fsType: ext4Params}
+		if fsType == "ext4" || fsType == "xfs" {
+			logrus.Infof("volume %v using user and longhorn provided %v fs creation params: %s", volume.Name, fsType, params)
+			cmdParamMapping := map[string]string{"mkfs." + fsType: params}
 			mounter = &mount.SafeFormatAndMount{
 				Interface: mount.New(""),
 				Exec:      NewForcedParamsExec(cmdParamMapping),
 			}
+		} else {
+			logrus.Infof("volume %v with unsupported filesystem %v, use default fs creation params", volume.Name, fsType)
```

Comment:
Should we use `Errorf` instead? also return an error if the fs not supported?

Reply:
Use `Warnf`, instead. Thank you.